ls1intum/Artemis
# `Exam mode`: Fix loading of result overview in  the summary of test runs

### Checklist
#### General
#### Server

### Motivation and Context
While working on [Exam mode: Fix loading of modeling exercise results in the summary of test runs #7590](https://github.com/ls1intum/Artemis/pull/7590)
I've noticed that the summary overview is not loaded properly.

This means that instructors have an experience that is further away from the student than needed when executing exam test runs.

### Description
* I have adjusted the database query `StudentExamRepository#findWithExercisesByUserIdAndExamId` to do not by default exclude all test run exams,
which introduced in [Improve performance in assessment dashboard database queries (#3081)](https://github.com/ls1intum/Artemis/pull/3081).
* Fixed the imports for tests according to our policy _(no wildcard imports)_
* Reduced code duplication in client for checking exam release date for results 
* Added client and server tests

### Steps for Testing
Prerequisites:
- 1 Instructor
- 1 Student
- 1 Ended Exam with a student participation

1. `Instructor`: Create and participate in an exam test run
2. `Instructor`: Assess at least one exercise in the test run
3. `Instructor`: View the summary of the test run 
4. `Instructor`: See that the summary table is loaded _(and no errors regarding the endpoint `/courses/{courseId}/exams/{examId}/student-exams/grade-summary` are displayed in the network tab)_
5. `Instructor`: Verify that you can still view student exams of ended exams
6. `Student`: Verify the exma summary is still displayed properly

### Testserver States
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test1)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test2)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test3)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test4)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test5)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test6)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test9)

### Review Progress
#### Performance Review
#### Code Review
#### Manual Tests

### Test Coverage

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------------:|
| [exam-participation.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5759/latest/artifact/shared/Coverage-Report-Client-Tests/app/exam/participate/exam-participation.service.ts.html) | 77.95% |                           ✅ |
| [exam-result-summary.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5759/latest/artifact/shared/Coverage-Report-Client-Tests/app/exam/participate/summary/exam-result-summary.component.ts.html) | 91.35% |                          ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------------:|
| [StudentExamRepository.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5759/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.repository/StudentExamRepository.html) | 96% |                            ✅ |
| [ExamRegistrationService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5759/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.exam/ExamRegistrationService.html) | 89% |                            ✅ |
| [ExamService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5759/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.exam/ExamService.html) | 93% |                            ✅ |
| [ExamSubmissionService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5759/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.exam/ExamSubmissionService.html) | 100% |                            |
| [StudentExamResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5759/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/StudentExamResource.html) | 86% |                          ✅ |

### Screenshots
#### New
The summary table is displayed (and the grade and grading key if the test run was assessed and a reading key is defined)

![Picture](https://github.com/ls1intum/Artemis/assets/63976129/5ee2b694-726e-496f-a7a4-2a3f84df702e)

#### Old
* Result Summary section is not displayed as it cannot be loaded
* error message in the console

```json
{
    "headers": {
        "normalizedNames": {},
        "lazyUpdate": null
    },
    "status": 404,
    "statusText": "Not Found",
    "url": "http://localhost:9000/api/courses/279/exams/10/student-exams/grade-summary?userId=1",
    "ok": false,
    "name": "HttpErrorResponse",
    "message": "Http failure response for http://localhost:9000/api/courses/279/exams/10/student-exams/grade-summary?userId=1: 404 Not Found",
    "error": {
        "type": "https://www.jhipster.tech/problem/problem-with-message",
        "status": 404,
        "path": "/api/courses/279/exams/10/student-exams/grade-summary",
        "message": "Requested resource does not exist."
    }
}
```

![Picture](https://github.com/ls1intum/Artemis/assets/63976129/d317e4dc-f40e-4d66-a177-90663281a7b1)
